### ISSUE TITLE
go mod vendor broken 

### ISSUE THREAD

**Issue by benma** [ethereum/go-ethereum#20232]
This package is not `go mod vendor` compatible due to this issue: https://github.com/golang/go/issues/26366#issuecomment-405683150

The problem is that `go mod vendor` does not vendor `libsecp256k1/include/secp256k1.h` referenced by `crypto/secp256k1/curve.go`.

```Go
/*
#include "libsecp256k1/include/secp256k1.h"
extern int secp256k1_ext_scalar_mul(const secp256k1_context* ctx, const unsigned char *point, const unsigned char *scalar);
*/
import "C"
```

This results in a build error when building with `go build -mod=vendor`:

```
vendor/github.com/ethereum/go-ethereum/crypto/secp256k1/curve.go:42:10: fatal error: libsecp256k1/include/secp256k1.h: No such file or directory
 #include "libsecp256k1/include/secp256k1.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

A similar issue exists in karalabe/usb, which is a dependency of go-ethereum, so both would need to be fixed: https://github.com/karalabe/usb/issues/13

**Comment by karalabe** [comment]
Yes, this is a limitation of Go's builders. I've opened a lot of issues upstream, they are always closed as `wontfix`. The only meaningful alternative is to manually vendor in the C files. Unfortunately there's not much we can do.

On the upside, AFAIK it should work **without** vendoring (go mod downloads the entire repo), so anyone using your stuff with `go mod` should be fine, only the vendoring workflow is messy, which is more of a burden for maintainers to keep clean and less for users.

**Comment by benma** [comment]
> On the upside, AFAIK it should work without vendoring (go mod downloads the entire repo), so anyone using your stuff with go mod should be fine, only the vendoring workflow is messy, which is more of a burden for maintainers to keep clean and less for users.

For better or worse, I need to use vendoring (`go mod vendor`). 

The referenced [issue comment](https://github.com/golang/go/issues/26366#issuecomment-405683150) suggests that it is fixable (by the module author, not the module user):

> Sorry, but the model for go builds is that what's needed to build a package is in that directory, not subdirectories. The way gen2brain/aac-go is structured, modifying the C source code in the subdirectories will not trigger a rebuild. Instead your build will use stale cache entries. The C code needs to be moved up to the package directory to make it build properly, not to mention vendor properly.

A way to [fix it](https://github.com/golang/go/issues/26366#issuecomment-405026625):

> One way to solve the cgo problem is to require a dummy go file in the c folder that is then imported by another package that uses the c files with a "_" import.

Fixing this means your library users can use your modules with standard go tooling, which would be nice.

**Comment by karalabe** [comment]
> One way to solve the cgo problem is to require a dummy go file in the c folder that is then imported by another package that uses the c files with a "_" import.

That would require the folder containing all the `.c` and `.h` files to be buildable and linkable together in the dumbest way without any preprocessing or exclusions, which generally cannot. Feel free to try and open a PR. Would be very happy to make our code `go mod` friendly though.

**Comment by yellowred** [comment]
There is still an error when I run tests with `mod=vendor`:
```
# github.com/karalabe/usb
vendor/github.com/karalabe/usb/hid_enabled.go:22:10: fatal error: './hidapi/hidapi/hidapi.h' file not found
#include "./hidapi/hidapi/hidapi.h"
```
Guess `usb` package needs a dummy go file as well.

**Comment by samu-gataca** [comment]
Hi all,

same problem here.

In past versions, I had the error mentioned above
`
vendor/github.com/ethereum/go-ethereum/crypto/secp256k1/curve.go:42:10: fatal error: libsecp256k1/include/secp256k1.h: No such file or directory
 #include "libsecp256k1/include/secp256k1.h"
`
Resolved, with this solution ... I know, it's not the best. But it works.
`
cp -r "${GOPATH}/src/github.com/ethereum/go-ethereum/crypto/secp256k1/libsecp256k1" $(pwd)/vendor/github.com/ethereum/go-ethereum/crypto/secp256k1/
`

Now, I tried to upgrade go-ethereum module to latest one, but in the compiling phase, we are having the new problem
`
# github.com/karalabe/usb
vendor/github.com/karalabe/usb/hid_enabled.go:22:10: fatal error: './hidapi/hidapi/hidapi.h' file not found
#include "./hidapi/hidapi/hidapi.h"
`
Do you have any temporary solution? Do you know any approximate date to solve the problem?

Regards

**Comment by fjl** [comment]
This is fixed in https://github.com/karalabe/usb/pull/17

**Comment by samu-gataca** [comment]
> This is fixed by [karalabe/usb#17](https://github.com/karalabe/usb/pull/17)

And then, it needs to be upgraded by go-ethereum :(
In the meantime, I'm using [Vend](https://github.com/nomad-software/vend), to use instead of go mod vendor, for that own project, using go-ethereum as a client.
